### ISSUE TITLE
Source code of zulu-openjdk?

### ISSUE THREAD

**Issue by pombredanne** [zulu-openjdk/zulu-openjdk#263]
Would you know where I can download the source code of your alpine builds?
Thanks!

**Comment by giltene** [comment]
TLDR: Zulu binaries *are* real, 100% Open Source. It is diligently packaged as such. Unlike many other binary packages that are not quite that.

"Open Source" is a very well-defined term. That definition has long been documented by the OSI, at opensource.org, under ["Open Source Definition"](https://opensource.org/osd). As someone who has been producing and contributing Open Source software for well over 30 years, and since well before this definition was solidified, I can assure you that Zulu carefully adheres to both the Open Source definition and to open source principles. It also carefully adheres to the various OSS licenses associated with the various parts that make up the overall Zulu package. We take this adherence very seriously. More so than most. That adherence comes with a fairly large amount of work under the covers since Zulu distributes (literally) hundreds of new binary artifacts per quarter, requiring careful tracking of each to its specific source code in order to be able to *actually* comply with OSS source code availability requirements as spelled out in the *actual* licenses involved.

Contrary to the [not so popular] belief held by some, "posting some version of the source code for some specific package variant on GitHub" does nothing to satisfy many actual OSS license requirements, including several that are part of the OpenJDK code base. A common theme in those OSS license is the requirement for the source code being shared to be the actual source code used to build the actual binary in question.

For a specific example, [GPLv2](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html) (under which large parts of OpenJDK are licensed) includes very specific requirements for making source code available. Under section 3, which describes the source code sharing requirements for distributing binaries under GPLv2, the license lists only 3 ways by which one can comply with the source-code availability requirements when distributing binaries. "Look up the source on GitHub by version" is not one of them, and doing so does not relieve one of the burden of adhering by the actual GPLv2 requirements.

IANAL, but to my understanding, wherever you find a binary that is (at least in part) licensed under GPLv2, and none of the 3 specific means of satisfying the source code availability requirements is available to you as the holder of that binary, you are looking at an OSS license violation. Those are unfortunately quite common, including with many binary artifacts that can be found here on GitHub, and whose as-is re-distribution would plainly violate GPLv2 requirements as some downstream receiver of the binary has no way by which to request or locate the exact source code used to build the binary they have...

Zulu binary packages carefully avoid this pitfall by transmitting the information needed to receive all required source code within the actual binary package. This satisfies e.g. section 3.b of GPLv2, as well as similar requirements in other OSS licenses related to Zulu contents.

There is nothing hard about getting Zulu source code. The instructions for doing so sit right there in the readme.txt in the base directory of each binary. And yes, the details can easily differ from binary to binary since the specific ID codes needed for each binary artifact may vary, even within a given quarterly update release.

**Comment by pombredanne** [comment]
@giltene Thanks for the details, but where are the sources? Your detailed answer fees like gaslighting.

If the source are not readily available, and I need to send an email where is the "open" part? Yes, you are complying to the letter, but not the spirit. You could comply equally well by making the sources available for download with the binaries, as any bona-fide open source projects does. 

The readme.txt you mention states:

>  Certain portions of this software are based on source code from OpenJDK
> (http://openjdk.java.net/)  and  licensed  under  the GNU General Public
> License  version  2  (GPLv2)  with   the  Classpath  Exception  (http://
> openjdk.java.net/legal/gplv2+ce.html).  For a period of three years from
> the date  of your receipt  of  this  software,  Azul  will  provide upon
> request, a complete  machine readable  copy of the  source code for such
> portions  based  on  OpenJDK on a medium  customarily used  for software
> interchange for a charge no more  than the cost of physically performing
> source distribution.
> 
> 
>   Please email azul_openjdk@azul.com for further information.
> 
>   Include this version code in your email:
>   Zulu 8.76.0.17 6bcb0574-a65d-4b51-ba80-0b8a8ca8fcac
> 
> To learn more about Azul products, visit https://www.azul.com/modern-cloud-enterprise/?r=zulu


From what I can, this is only about "certain portion"?

**Comment by pombredanne** [comment]
I emailed azul_openjdk@azul.com:

> Hi:
> Is there a place to download the source code of a JDK build of yours?
> Thanks!

And I received this reply from a Pavel Petroshenko from Azul:

> Hi Philippe,
> 
> Zulu sources are available to anyone with a binary looking for the corresponding source bundle. Every Zulu configuration is different and thus may have its own unique source code. To identify the sources, we need the binary bundle identifier, which is available in the readme file. Please follow the instructions in the file, and we are happy to assist you with the sources. We are looking into simplifying the process for everyone, but it is what it is now.
> 
> So, can you please follow the instructions from the readme.txt files located in the binary images and share the source code IDs?
> 
> Does that make sense?
> 
> Thank you!
> Pavel

**Comment by pombredanne** [comment]
You do not distribute the binaries by email request, why distribute the source by email? This feels completely nuts.

**Comment by captn3m0** [comment]
I sent a source request to `azul_openjdk@azul.com` (`zulu22.30.13-ca-jdk22.0.1-linux_x64.zip`  / `
Zulu 22.30+13 995dc5ae-77a7-4214-80fb-da03dc88f946`) and got a prompt response with a download link to a zip file with the source code, duly licensed.

The `.branding` file contains:

```
JDK_TAG=jdk-22.0.1+8
JDK_VERSION_DATE=2024-04-16

ZULU_TAG=zulu-22.30+13

ZULU_FEATURE_NUMBER=22
ZULU_UPDATE_NUMBER=30
ZULU_BUILD_NUMBER=13
```

The diff against the corresponding [openJDK build](https://codeload.github.com/openjdk/jdk22u/zip/refs/tags/jdk-22.0.1%2B8) is [fairly minimal](https://gist.github.com/captn3m0/a7b9589266da65d7c19ba2c01901abcb).
